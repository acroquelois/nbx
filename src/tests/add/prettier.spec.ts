import { filesystem, system } from 'gluegun';
const stripANSI = require('strip-ansi');
import * as prompts from 'prompts';
import * as latestVersion from 'latest-version';
import Prettier from '../../commands/add/prettier';

jest.setTimeout(30000);

/* eslint-disable no-console,max-nested-callbacks,@typescript-eslint/ban-ts-ignore */
let consoleLogOutput: string[];
let execPath: string;
const originalCwd = process.cwd();
const originalLog = console.log;

beforeEach(async () => {
  consoleLogOutput = [];
  console.log = (x: any) => consoleLogOutput.push(stripANSI(x));

  const tmpDirName = `${new Date().getTime()}-${Math.random() * 100}-add-nbx-test`;
  execPath = filesystem.path('/', 'tmp', tmpDirName);
  filesystem.dir(execPath);
  process.chdir(execPath);
});
afterEach(() => {
  console.log = originalLog;
  process.chdir(originalCwd);
  jest.restoreAllMocks();
  filesystem.remove(execPath);
});

const initWithConfigAndCommit = async () => {
  await system.run('git init');
  filesystem.write('.nbxrc', { git: { user: 'aaa', email: 'bbb' } });
  filesystem.write('package.json', {});
  filesystem.write('test.html', '<html><body><a>testokiii</a></body></html>');
  await system.run('touch yarn.lock');
  await system.run('echo node_modules > .gitignore');
  await system.run('git add * .nbxrc .gitignore');
  await system.run('git commit -m "init state" --author="Dummy <dummy@noop.noop>');
};

describe('prettier', () => {
  it('should print help correctly', async () => {
    try {
      await Prettier.run(['-h']);
    } catch {}
    expect(consoleLogOutput).toMatchSnapshot();
  });

  it('should work without commits', async () => {
    await initWithConfigAndCommit();
    prompts.inject(['**/*.{js,vue,json,ts,tsx,md,yml,html}', false]);
    const [latestHusky, latestPrettier, latestPrettyQuick] = await Promise.all([
      latestVersion('husky'),
      latestVersion('prettier'),
      latestVersion('pretty-quick'),
    ]);

    await Prettier.run(['--no-spinner']);

    const packageJson = filesystem.read('package.json', 'json');
    const htmlFile = filesystem.read('test.html', 'utf8');
    expect(consoleLogOutput).toMatchSnapshot();
    expect(packageJson).toStrictEqual({
      devDependencies: {
        'husky': latestHusky,
        'prettier': latestPrettier,
        'pretty-quick': latestPrettyQuick,
      },
      husky: {
        hooks: {
          'pre-commit': 'pretty-quick --staged',
        },
      },
      scripts: {
        'format:check': 'prettier --list-different "**/*.{js,vue,json,ts,tsx,md,yml,html}"',
        'format:write': 'prettier --write "**/*.{js,vue,json,ts,tsx,md,yml,html}"',
      },
    });
    expect(htmlFile).toMatchSnapshot();
    const after = await system.run('git log --name-status  --format="%s"');
    const afterSlitted = after
      .split('\n')
      .map(val => val.trim())
      .map(val => stripANSI(val))
      .filter(val => val !== '');

    expect(afterSlitted).toHaveLength(6);
  });
  it('should work with commits', async () => {
    await initWithConfigAndCommit();
    prompts.inject(['**/*.{js,vue,json,ts,tsx,md,yml,html}', true]);
    const [latestHusky, latestPrettier, latestPrettyQuick] = await Promise.all([
      latestVersion('husky'),
      latestVersion('prettier'),
      latestVersion('pretty-quick'),
    ]);

    await Prettier.run(['--no-spinner']);

    const packageJson = filesystem.read('package.json', 'json');
    const htmlFile = filesystem.read('test.html', 'utf8');
    expect(consoleLogOutput).toMatchSnapshot();
    expect(packageJson).toStrictEqual({
      devDependencies: {
        'husky': latestHusky,
        'prettier': latestPrettier,
        'pretty-quick': latestPrettyQuick,
      },
      husky: {
        hooks: {
          'pre-commit': 'pretty-quick --staged',
        },
      },
      scripts: {
        'format:check': 'prettier --list-different "**/*.{js,vue,json,ts,tsx,md,yml,html}"',
        'format:write': 'prettier --write "**/*.{js,vue,json,ts,tsx,md,yml,html}"',
      },
    });
    expect(htmlFile).toMatchSnapshot();
    const after = await system.run('git log --name-status  --format="%s"');
    const afterSlitted = after
      .split('\n')
      .map(val => val.trim())
      .map(val => stripANSI(val))
      .filter(val => val !== '');

    expect(afterSlitted).toStrictEqual([
      ':art: apply prettier style to project',
      'M\tpackage.json',
      'M\ttest.html',
      ':wrench: add prettierrc config file',
      'A\t.prettierrc',
      ':wrench: add script and husky to package.json',
      'M\tpackage.json',
      `:heavy_plus_sign: add pretty-quick@${latestPrettyQuick} as a dev dependency`,
      'M\tpackage.json',
      'M\tyarn.lock',
      `:heavy_plus_sign: add husky@${latestHusky} as a dev dependency`,
      'M\tpackage.json',
      'M\tyarn.lock',
      `:heavy_plus_sign: add prettier@${latestPrettier} as a dev dependency`,
      'M\tpackage.json',
      'M\tyarn.lock',
      'init state',
      'A\t.gitignore',
      'A\t.nbxrc',
      'A\tpackage.json',
      'A\ttest.html',
      'A\tyarn.lock',
    ]);
  });
});
