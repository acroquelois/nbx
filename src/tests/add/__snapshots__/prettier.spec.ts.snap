// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prettier should print help correctly 1`] = `
Array [
  "add prettier to project and format it
",
  "USAGE
  $ nbx add:prettier

OPTIONS
  -h, --help      show CLI help
  -v, --verbose   Verbose output
  --[no-]spinner  Enable spinner in cli output, true by default",
  "",
]
`;

exports[`prettier should work with commits 1`] = `
Array [
  "Adding prettier as a dev dependency",
  "The step was done without any error.",
  "Adding husky as a dev dependency",
  "The step was done without any error.",
  "Adding pretty-quick as a dev dependency",
  "The step was done without any error.",
  "Adding package.json scripts",
  "The step was done without any error.",
  "Adding .prettierrc config file",
  "The step was done without any error.",
  "Updating code to match prettier style",
  "The step was done without any error.",
]
`;

exports[`prettier should work with commits 2`] = `
"<html>
  <body>
    <a>testokiii</a>
  </body>
</html>
"
`;

exports[`prettier should work without commits 1`] = `
Array [
  "Adding prettier as a dev dependency",
  "The step was done without any error.",
  "Adding husky as a dev dependency",
  "The step was done without any error.",
  "Adding pretty-quick as a dev dependency",
  "The step was done without any error.",
  "Adding package.json scripts",
  "The step was done without any error.",
  "Adding .prettierrc config file",
  "The step was done without any error.",
  "Updating code to match prettier style",
  "The step was done without any error.",
]
`;

exports[`prettier should work without commits 2`] = `
"<html>
  <body>
    <a>testokiii</a>
  </body>
</html>
"
`;
